### PR TITLE
move Find* files to cogutil

### PR DESCRIPTION
**Pull request by noskill**
As part of the work on issue https://github.com/opencog/cogutil/issues/113 moving Find* cmake files to cogutil
depends on https://github.com/opencog/atomspace/pull/1844

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -55,6 +54,24 @@ ADD_DEFINITIONS(-DPROJECT_SOURCE_DIR=\\"${CMAKE_SOURCE_DIR}\\"
 # add the 'lib' dir to cmake's module search path
 SET(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/lib/")
```
Comment (vsbogd) [inline CMakeLists.txt]: Seems not needed anymore since line 58 is added:

    list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/lib/")

Comment (noskill) [reply CMakeLists.txt]: Ok

**Comment by vsbogd** [comment]
I can build OpenCog but cannot NLP unit tests:

/usr/bin/x86_64-linux-gnu-ld: cannot find -lcogutil
collect2: error: ld returned 1 exit status
tests/nlp/microplanning/CMakeFiles/MicroplanningUTest.dir/build.make:122: recipe for target 'tests/nlp/microplanning/MicroplanningUTest' failed
make[3]: *** [tests/nlp/microplanning/MicroplanningUTest] Error 1
CMakeFiles/Makefile2:3506: recipe for target 'tests/nlp/microplanning/CMakeFiles/MicroplanningUTest.dir/all' failed
make[2]: *** [tests/nlp/microplanning/CMakeFiles/MicroplanningUTest.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
[ 33%] Built target patternindex
[ 35%] Built target attention-types

**Comment by noskill** [comment]
@vsbogd My latest changes fixed problem, kind of.. On one of my machines I got linking error:

```
[ 49%] Linking CXX executable SuRealUTest
/tmp/test/lib/opencog/libquery.so: undefined reference to `opencog::EvaluationLink::do_eval_scratch(opencog::AtomSpace*, opencog::Handle const&, opencog::AtomSpace*, bool)'
```
Could you please check if you can  build tests?

**Comment by vsbogd** [comment]
Same problem but with another test 
```
[ 84%] Linking CXX executable PatternMinerUTest
/home/vital/projects/opencog/test/build/lib/opencog/libquery.so: undefined reference to `opencog::Instantiator::Instantiator(opencog::AtomSpace*)'
/home/vital/projects/opencog/test/build/lib/opencog/libquery.so: undefined reference to `opencog::Instantiator::instantiate(opencog::Handle const&, std::map<opencog::Handle, opencog::Handle, std::less<opencog::Handle>, std::allocator<std::pair<opencog::Handle const, opencog::Handle> > > const&, bool)'
/home/vital/projects/opencog/test/build/lib/opencog/libquery.so: undefined reference to `opencog::EvaluationLink::do_eval_scratch(opencog::AtomSpace*, opencog::Handle const&, opencog::AtomSpace*, bool)'
/home/vital/projects/opencog/test/build/lib/opencog/libquery.so: undefined reference to `opencog::EvaluationLink::do_evaluate(opencog::AtomSpace*, opencog::Handle const&, bool)'
collect2: error: ld returned 1 exit status
tests/learning/PatternMiner/CMakeFiles/PatternMinerUTest.dir/build.make:124: recipe for target 'tests/learning/PatternMiner/PatternMinerUTest' failed
make[3]: *** [tests/learning/PatternMiner/PatternMinerUTest] Error 1
CMakeFiles/Makefile2:3353: recipe for target 'tests/learning/PatternMiner/CMakeFiles/PatternMinerUTest.dir/all' failed
make[2]: *** [tests/learning/PatternMiner/CMakeFiles/PatternMinerUTest.dir/all] Error 2
CMakeFiles/Makefile2:151: recipe for target 'CMakeFiles/tests.dir/rule' failed
make[1]: *** [CMakeFiles/tests.dir/rule] Error 2
Makefile:212: recipe for target 'tests' failed
make: *** [tests] Error 2
```

**Comment by vsbogd** [comment]
Looks strange because ```gcc``` command line contains libexecution.so which exports required symbols:
```
/usr/bin/g++  -Wall -fPIC -Wno-variadic-macros -fopenmp -std=gnu++11 -O2 -g -fstack-protector  -rdynamic CMakeFiles/PatternMinerUTest.dir/PatternMinerUTest.cpp.o  -o PatternMinerUTest -Wl,-rpath,/home/vital/projects/opencog/test/opencog-build/opencog/learning/PatternMiner:/home/vital/projects/opencog/test/opencog-build/opencog/learning/PatternMiner/types:/home/vital/projects/opencog/test/build/lib/opencog:/home/vital/projects/opencog/test/build/lib ../../../opencog/learning/PatternMiner/libPatternMiner.so ../../../opencog/learning/PatternMiner/types/libpatternminer-types.so /home/vital/projects/opencog/test/build/lib/opencog/libattentionbank.so /home/vital/projects/opencog/test/build/lib/opencog/libpersist.so /home/vital/projects/opencog/test/build/lib/opencog/libruleengine.so /home/vital/projects/opencog/test/build/lib/opencog/libatomspaceutils.so /home/vital/projects/opencog/test/build/lib/opencog/libexecution.so /home/vital/projects/opencog/test/build/lib/opencog/libquery.so /home/vital/projects/opencog/test/build/lib/opencog/libPythonEval.so -lboost_filesystem -lpython3.6m /home/vital/projects/opencog/test/build/lib/opencog/libsmob.so /home/vital/projects/opencog/test/build/lib/opencog/libclearbox.so -Wl,--no-as-needed -lguile-2.2 /home/vital/projects/opencog/test/build/lib/opencog/libunify.so /home/vital/projects/opencog/test/build/lib/opencog/liblambda.so /home/vital/projects/opencog/test/build/lib/opencog/libatomspace.so /home/vital/projects/opencog/test/build/lib/opencog/libatomcore.so /home/vital/projects/opencog/test/build/lib/opencog/libatomutils.so /home/vital/projects/opencog/test/build/lib/opencog/libatombase.so /home/vital/projects/opencog/test/build/lib/opencog/libatomproto.so /home/vital/projects/opencog/test/build/lib/opencog/libtruthvalue.so -lboost_thread -lpthread /home/vital/projects/opencog/test/build/lib/libcogutil.so -lboost_system -lboost_regex
/home/vital/projects/opencog/test/build/lib/opencog/libquery.so: undefined reference to `opencog::Instantiator::Instantiator(opencog::AtomSpace*)'
/home/vital/projects/opencog/test/build/lib/opencog/libquery.so: undefined reference to `opencog::Instantiator::instantiate(opencog::Handle const&, std::map<opencog::Handle, opencog::Handle, std::less<opencog::Handle>, std::allocator<std::pair<opencog::Handle const, opencog::Handle> > > const&, bool)'
/home/vital/projects/opencog/test/build/lib/opencog/libquery.so: undefined reference to `opencog::EvaluationLink::do_eval_scratch(opencog::AtomSpace*, opencog::Handle const&, opencog::AtomSpace*, bool)'
/home/vital/projects/opencog/test/build/lib/opencog/libquery.so: undefined reference to `opencog::EvaluationLink::do_evaluate(opencog::AtomSpace*, opencog::Handle const&, bool)'
collect2: error: ld returned 1 exit status
```

**Comment by noskill** [comment]
@vsbogd No errors when building on opensuse with the same flags.. But on ubuntu i found that if I add  "-Wl,--no-as-needed <path to libexecution.so>" to g++ flags then test links without error, still struggling with cmake to do it..

**Comment by noskill** [comment]
@vsbogd I fixed linking problem in this commit https://github.com/opencog/atomspace/pull/1844/commits/c2deabbbc6595b09718b93d791fcfca8d5b5064d by setting IMPORTED_LINK_INTERFACE_LIBRARIES_RELEASE property of libquery. cmake is not very good if you have circular dependencies, don't know if there is a better solution.

**Comment by vsbogd** [comment]
Made clean build, built tests. 
Test results on my machine (501d8711e6612d94120018fe1547468aa880bd2a):
```
50% tests passed, 8 tests failed out of 16

Total Test time (real) =   7.96 sec

The following tests FAILED:
	  2 - HebbianCreationModuleUTest (Failed)
	  3 - CogServerUTest (SEGFAULT)
	  6 - AnaphoraTest (Failed)
	  7 - MOSESPLNSynergyUTest (Failed)
	  9 - OpenPsiRulesUTest (Failed)
	 10 - OpenPsiImplicatorUTest (Failed)
	 11 - OpenPsiSCMUTest (Failed)
	 16 - MinerUTest (Failed)
```

comparing to master (890a8219fe2f590bcc7959274967f3f55d09b01a):
```
50% tests passed, 9 tests failed out of 18

Total Test time (real) =  24.73 sec

The following tests FAILED:
	  3 - CogServerUTest (SEGFAULT)
	  6 - AnaphoraTest (Failed)
	  7 - SuRealUTest (Failed)
	  8 - MicroplanningUTest (Failed)
	  9 - MOSESPLNSynergyUTest (Failed)
	 11 - OpenPsiRulesUTest (Failed)
	 12 - OpenPsiImplicatorUTest (Failed)
	 13 - OpenPsiSCMUTest (Failed)
	 18 - MinerUTest (Failed)
```

**Comment by linas** [comment]
why does master have more unit tests?  Why do you have fewer unit tests than I? I have 20 unit tests:
```
      Start  1: AttentionParamQueryUTest
 1/20 Test  #1: AttentionParamQueryUTest .........   Passed    0.12 sec
      Start  2: HebbianCreationModuleUTest
 2/20 Test  #2: HebbianCreationModuleUTest .......   Passed    3.55 sec
      Start  3: CogServerUTest
 3/20 Test  #3: CogServerUTest ...................   Passed    1.51 sec
      Start  4: ShellUTest
 4/20 Test  #4: ShellUTest .......................   Passed  123.75 sec
      Start  5: AgentUTest
 5/20 Test  #5: AgentUTest .......................   Passed    0.77 sec
      Start  6: AtomOcTreeUTest
 6/20 Test  #6: AtomOcTreeUTest ..................   Passed    0.17 sec
      Start  7: TimeSpaceAtomUTest
 7/20 Test  #7: TimeSpaceAtomUTest ...............   Passed    0.27 sec
      Start  8: AnaphoraTest
 8/20 Test  #8: AnaphoraTest .....................***Failed    4.44 sec
      Start  9: SuRealUTest
 9/20 Test  #9: SuRealUTest ......................   Passed   43.54 sec
      Start 10: MicroplanningUTest
10/20 Test #10: MicroplanningUTest ...............   Passed   55.79 sec
      Start 11: MOSESPLNSynergyUTest
11/20 Test #11: MOSESPLNSynergyUTest .............   Passed   12.96 sec
      Start 12: PLNRulesUTest
12/20 Test #12: PLNRulesUTest ....................   Passed   10.12 sec
      Start 13: OpenPsiRulesUTest
13/20 Test #13: OpenPsiRulesUTest ................   Passed    1.78 sec
      Start 14: OpenPsiImplicatorUTest
14/20 Test #14: OpenPsiImplicatorUTest ...........   Passed    1.74 sec
      Start 15: OpenPsiSCMUTest
15/20 Test #15: OpenPsiSCMUTest ..................   Passed    0.84 sec
      Start 16: DimEmbedUTest
16/20 Test #16: DimEmbedUTest ....................   Passed    0.84 sec
      Start 17: StatisticsUTest
17/20 Test #17: StatisticsUTest ..................   Passed    0.09 sec
      Start 18: PatternMinerUTest
18/20 Test #18: PatternMinerUTest ................   Passed    1.11 sec
      Start 19: PatternIndexAPIUTest
19/20 Test #19: PatternIndexAPIUTest .............   Passed    4.53 sec
      Start 20: MinerUTest
20/20 Test #20: MinerUTest .......................   Passed   21.94 sec

95% tests passed, 1 tests failed out of 20

Total Test time (real) = 289.90 sec

The following tests FAILED:
	  8 - AnaphoraTest (Failed)
Errors while running CTest
```

**Comment by vsbogd** [comment]
I have added dependencies to make OpenCog SpaceTime. Now I have 21 unit test, CogServerUTest is passed now, but I have to skip ShellUTest as it hangs. HebbianCreationModuleUTest passed after ```make install```/

Overall picture for master (43ea0858f640a2230a87872938451bc0c3da6e60):
```
60% tests passed, 8 tests failed out of 20

Total Test time (real) =  26.10 sec

The following tests FAILED:
	  8 - AnaphoraTest (Failed)
	  9 - SuRealUTest (Failed)
	 10 - MicroplanningUTest (Failed)
	 11 - MOSESPLNSynergyUTest (Failed)
	 13 - OpenPsiRulesUTest (Failed)
	 14 - OpenPsiImplicatorUTest (Failed)
	 15 - OpenPsiSCMUTest (Failed)
	 20 - MinerUTest (Failed)
```

**Comment by vsbogd** [comment]
Thanks, I have checked my version, it has ```-q``` parameter, so it is probably a root cause.

**Comment by AmeBel** [comment]
@vsbogd you mention you have 21 unit tests, I have only 20. Would you please list the 21 test, for comparison?

**Comment by vsbogd** [comment]
@linas , @ngeiswei, I think it is ready to be merged, if you don't have any objections I would merge it.

**Comment by ngeiswei** [comment]
I didn't look at it but I trust your judgment, @vsbogd. Just wait for @linas' approval.

**Comment by noskill** [comment]
@ngeiswei We decided to do some more testing, please do not merge yet.

File: CMakeLists.txt [hunk 2]
```diff
@@ -246,7 +249,7 @@ ENDIF (GTK3_FOUND)
 
 # Gnu Guile scheme interpreter
 # Version 2.2.2 is needed for mrs io ports, compilation, atomics, nlp
-FIND_PACKAGE(Guile 2.2.2)
+FIND_PACKAGE(Guile 2.0.2)
 IF (GUILE_FOUND AND GMP_LIBRARY AND GMP_INCLUDE_DIR)
```
Comment (linas) [inline CMakeLists.txt]: why are you down-grading guile? Like the comment up above says: `Version 2.2.2 is needed for mrs io ports, compilation, atomics, nlp`

**Comment by linas** [comment]
Per my most recent comment on #113 I'm starting to feel a bit nervous/wishy-washy about this. 

Moving some of the files makes sense: for example, atomspace and opencog and agi-bio have some files that are cut-n-pases of one-another. Moving those to a common location makes sense.

Other files are not shared, and probably will never be: no one, except for opencog, will ever need `FindLinkGrammar.cmake`, so moving it to the far-away land of cogutils seems to be a bit extreme.

**Comment by vsbogd** [comment]
Just to make it more clear: #113 above means https://github.com/opencog/cogutil/issues/113

**Comment by noskill** [comment]
@ngeiswei Ok, i merged recent changes from cogutil from https://github.com/opencog/cogutil/pull/123 , should ok now.

**Comment by noskill** [comment]
depends on https://github.com/opencog/moses/pull/60

**Comment by vsbogd** [comment]
Still retesting it after last commits:
- cogutil - ok
- atomspace - ok
- moses - to be done
- opencog - to be done

**Comment by vsbogd** [comment]
Finished testing:
- cogutil - ok
- atomspace - ok
- moses - ok
- opencog - ok

**Comment by vsbogd** [comment]
Thanks!

**Pull request merged**